File: 874-walking-robot-simulation/874. Walking Robot Simulation.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        # Directions: North, East, South, West
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        direction = 0  # Starts facing North
        
        # Set of obstacles for O(1) lookup
        obstacle_set = set(map(tuple, obstacles))
        
        x, y = 0, 0
        max_distance_sq = 0
        
        for command in commands:
            if command == -2:  # Turn left
                direction = (direction - 1) % 4
            elif command == -1:  # Turn right
                direction = (direction + 1) % 4
            else:  # Move forward
                for _ in range(command):
                    next_x = x + directions[direction][0]
                    next_y = y + directions[direction][1]
                    if (next_x, next_y) not in obstacle_set:
                        x, y = next_x, next_y
                        max_distance_sq = max(max_distance_sq, x * x + y * y)
                    else:
                        break
        
        return max_distance_sq

File: 874-walking-robot-simulation/874. Walking Robot Simulation.cpp
class Solution {
public:
    struct pair_hash {
        size_t operator()(const pair<int, int>& p) const {
            return hash<int>()(p.first) ^ (hash<int>()(p.second) << 1);
        }
    };
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        unordered_set<pair<int, int>, pair_hash> obstacles_set;
        vector<string> directions = {"N", "R", "S", "L"};
        int idx = 0;
        for (auto pair : obstacles) {
            obstacles_set.insert({pair[0], pair[1]});
        }
        int x = 0, y = 0;
        int res = 0;
        for (auto c : commands) {
            if (c == -2) {
                idx = (idx -1);
                if (idx == -1) idx = 3;
            } else if (c == -1) {
                idx = (idx + 1) % 4;
            } else {
                for (int i = 0; i < c; i++) {
                    cout << x << y << endl;
                    if (directions[idx] == "N") {
                        if (obstacles_set.count({x,y+1})) break;
                        y++;
                    } 
                    if (directions[idx] == "S") {
                        if (obstacles_set.count({x,y-1})) break;
                        y--;
                    } 
                    if (directions[idx] == "L") {
                        if (obstacles_set.count({x-1,y})) break;
                        x--;
                    } 
                    if (directions[idx] == "R") {
                        if (obstacles_set.count({x+1,y})) break;
                        x++;
                    } 
                    cout << x << y << endl;
                    res = max(res, x*x+y*y);
                }
            }
        }
        return res;
    }
};